src/sources/http.rs

```diff
@@ -0,0 +1,339 @@
+    net::SocketAddr,
+};
+use warp::http::{HeaderMap, StatusCode, HeaderValue};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct SimpleHttpConfig {
+    address: SocketAddr,
+    #[serde(default)]
+    encoding: Encoding,
+    #[serde(default)]
+    headers: Vec<String>,
+}
+
+#[derive(Clone)]
+struct SimpleHttpSource {
+    encoding: Encoding,
+    headers: Vec<String>,
+}
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Derivative, Copy)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+pub enum Encoding {
+    #[derivative(Default)]
+    Text,
+    Ndjson,
+    Json,
+}
+
+impl HttpSource for SimpleHttpSource {
+    fn build_event(
+        &self,
+        body: impl Buf,
+        header_map: HeaderMap,
+    ) -> Result<Vec<Event>, ErrorMessage> {
+        decode_body(body, self.encoding)
+            .map(|events| add_headers(events, &self.headers, header_map))
+    }
+}
+
+#[typetag::serde(name = "http")]
+impl SourceConfig for SimpleHttpConfig {
+    fn build(
+        &self,
+        _: &str,
+        _: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let source = SimpleHttpSource {
+            encoding: self.encoding.clone(),
+            headers: self.headers.clone(),
+        };
+        source.run(self.address, "".to_string(), out)
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "http"
+    }
+}
+
+fn add_headers(
+    mut events: Vec<Event>,
+    headers_config: &[String],
+    headers: HeaderMap,
+) -> Vec<Event> {
+    for header_name in headers_config {
+        let value = headers.get(header_name).map(HeaderValue::as_bytes).unwrap_or(b"");
+        for event in events.iter_mut() {
+            event.as_mut_log().insert(header_name as &str, value);
+        }
+    }
+
+    events
+}
+
+fn decode_body(body: impl Buf, enc: Encoding) -> Result<Vec<Event>, ErrorMessage> {
+    let mut rdr = BufReader::new(body.reader())
+        .lines()
+        .filter_map(|res| {
+            res.map_err(|error| error!(message = "Error reading request body", ?error))
+                .ok()
+        })
+        .filter(|s| !s.is_empty());
+
+    match enc {
+        Encoding::Text => Ok(rdr.map(Event::from).collect()),
+        Encoding::Ndjson => Ok(rdr
+            .map(|j| json_parse_object(json_parse(j)?))
+            .collect::<Result<_, _>>()?),
+        Encoding::Json => {
+            let first_line = rdr.next();
+            let second_line = rdr.next();
+            if first_line.is_some() && second_line.is_none() {
+                //need exactly one line
+                json_parse_array_of_object(json_parse(first_line.unwrap())?)
+            } else {
+                Err(json_error())
+            }
+        }
+    }
+}
+
+fn json_parse(json_str: String) -> Result<JsonValue, ErrorMessage> {
+    serde_json::from_str(&json_str).map_err(|_| json_error())
```

Comment:
I don't know if this function's purpose warrants it's existence. If all it does is throw away error context from `from_str` maybe we can just remove it entirely?